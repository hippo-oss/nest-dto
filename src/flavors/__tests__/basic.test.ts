import { plainToClass } from 'class-transformer';
import { validate } from 'class-validator';

import * as basic from '../basic';
import { INPUT, createFixtures } from './fixtures';

describe('flavors.basic', () => {
    const Example = createFixtures(basic);

    it('validates required fields', async () => {
        const obj = plainToClass(Example, {});

        // we do not expect any data to be transformed
        expect(Object.keys(obj)).toHaveLength(0);
        expect(obj).toMatchSnapshot();

        const errors = await validate(obj);
        // we expect an error for every required field (but not the optional ones)
        expect(errors).toHaveLength(14);
        expect(errors).toMatchSnapshot();
    });
    it('transforms input data', async () => {
        const obj = plainToClass(Example, INPUT);

        // expect all data to be transformed
        expect(Object.keys(obj)).toHaveLength(36);
        expect(obj).toMatchObject({
            falseBooleanValue: false,
            zeroBooleanValue: false,
            requiredObjectArrayValue: [{
                requiredStringValue: 'nested',
            }],
            requiredPrimitiveArrayValue: [
                42,
            ],
            requiredNestedValue: {
                requiredStringValue: 'nested',
            },
            optionalObjectArrayValue: [{
                optionalStringValue: 'nested',
            }],
            optionalPrimitiveArrayValue: [
                42,
            ],
            optionalNestedValue: {
                optionalStringValue: 'nested',
            },
            nullableObjectArrayValue: [{
                nullableStringValue: 'nested',
            }],
            nullablePrimitiveArrayValue: [
                42,
            ],
            nullableNestedValue: {
                nullableStringValue: 'nested',
            },
            // NB: This field is not transformed (b/c @Expose() is not included)
            string_value_with_different_wire_name: 'value',
        });

        const errors = await validate(obj);
        // we expect no errors, except for the field that was meant to be Expose'd
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not transform input data if extraneous values are excluded', async () => {
        const obj = plainToClass(Example, INPUT, {
            excludeExtraneousValues: true,
        });

        // expect no data to be transformed (b/c @Expose() is not included)
        expect(Object.keys(obj)).toHaveLength(0);

        const errors = await validate(obj);
        // we expect an error for every required field (but not the optional ones)
        expect(errors).toHaveLength(14);
        expect(errors).toMatchSnapshot();
    });
});
