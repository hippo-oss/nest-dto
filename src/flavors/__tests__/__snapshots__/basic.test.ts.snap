// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flavors.basic does not transform input data if extraneous values are excluded 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isArray": "requiredObjectArrayValue must be an array",
      "isDefined": "requiredObjectArrayValue should not be null or undefined",
    },
    "property": "requiredObjectArrayValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isArray": "requiredPrimitiveArrayValue must be an array",
      "isDefined": "requiredPrimitiveArrayValue should not be null or undefined",
    },
    "property": "requiredPrimitiveArrayValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "falseBooleanValue must be a boolean value",
      "isDefined": "falseBooleanValue should not be null or undefined",
    },
    "property": "falseBooleanValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "zeroBooleanValue must be a boolean value",
      "isDefined": "zeroBooleanValue should not be null or undefined",
    },
    "property": "zeroBooleanValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "requiredBooleanValue must be a boolean value",
      "isDefined": "requiredBooleanValue should not be null or undefined",
    },
    "property": "requiredBooleanValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredDateStringValue should not be null or undefined",
      "isIso8601": "requiredDateStringValue must be a valid ISO 8601 date string",
    },
    "property": "requiredDateStringValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDate": "requiredDateValue must be a Date instance",
      "isDefined": "requiredDateValue should not be null or undefined",
    },
    "property": "requiredDateValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredEnumValue should not be null or undefined",
      "isEnum": "requiredEnumValue must be a valid enum value",
    },
    "property": "requiredEnumValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredIntegerValue should not be null or undefined",
      "isInt": "requiredIntegerValue must be an integer number",
    },
    "property": "requiredIntegerValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredNestedValue should not be null or undefined",
    },
    "property": "requiredNestedValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredNumberValue should not be null or undefined",
      "isNumber": "requiredNumberValue must be a number conforming to the specified constraints",
    },
    "property": "requiredNumberValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredStringValue should not be null or undefined",
      "isString": "requiredStringValue must be a string",
    },
    "property": "requiredStringValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredUUIDValue should not be null or undefined",
      "isUuid": "requiredUUIDValue must be a UUID",
    },
    "property": "requiredUUIDValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "stringValueWithDifferentWireName should not be null or undefined",
      "isString": "stringValueWithDifferentWireName must be a string",
    },
    "property": "stringValueWithDifferentWireName",
    "target": Example {},
    "value": undefined,
  },
]
`;

exports[`flavors.basic transforms input data 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "stringValueWithDifferentWireName should not be null or undefined",
      "isString": "stringValueWithDifferentWireName must be a string",
    },
    "property": "stringValueWithDifferentWireName",
    "target": Example {
      "falseBooleanValue": false,
      "nullableBooleanValue": true,
      "nullableDateStringValue": "2021-01-12",
      "nullableDateValue": 2021-01-12T01:12:38.956Z,
      "nullableEnumValue": "Value",
      "nullableIntegerValue": 42,
      "nullableNestedValue": NestedNullableExample {
        "nullableStringValue": "nested",
      },
      "nullableNumberValue": 42,
      "nullableObjectArrayValue": Array [
        NestedNullableExample {
          "nullableStringValue": "nested",
        },
      ],
      "nullablePrimitiveArrayValue": Array [
        42,
      ],
      "nullableStringValue": "value",
      "nullableUUIDValue": "3430cef4-6e7a-43da-a5a7-ae4c6a18be47",
      "optionalBooleanValue": true,
      "optionalDateStringValue": "2021-01-12",
      "optionalDateValue": 2021-01-12T01:12:38.956Z,
      "optionalEnumValue": "Value",
      "optionalIntegerValue": 42,
      "optionalNestedValue": NestedOptionalExample {
        "optionalStringValue": "nested",
      },
      "optionalNumberValue": 42,
      "optionalObjectArrayValue": Array [
        NestedOptionalExample {
          "optionalStringValue": "nested",
        },
      ],
      "optionalPrimitiveArrayValue": Array [
        42,
      ],
      "optionalStringValue": "value",
      "optionalUUIDValue": "3430cef4-6e7a-43da-a5a7-ae4c6a18be47",
      "requiredBooleanValue": true,
      "requiredDateStringValue": "2021-01-12",
      "requiredDateValue": 2021-01-12T01:12:38.956Z,
      "requiredEnumValue": "Value",
      "requiredIntegerValue": 42,
      "requiredNestedValue": NestedRequiredExample {
        "requiredStringValue": "nested",
      },
      "requiredNumberValue": 42,
      "requiredObjectArrayValue": Array [
        NestedRequiredExample {
          "requiredStringValue": "nested",
        },
      ],
      "requiredPrimitiveArrayValue": Array [
        42,
      ],
      "requiredStringValue": "value",
      "requiredUUIDValue": "3430cef4-6e7a-43da-a5a7-ae4c6a18be47",
      "string_value_with_different_wire_name": "value",
      "zeroBooleanValue": false,
    },
    "value": undefined,
  },
]
`;

exports[`flavors.basic validates required fields 1`] = `Example {}`;

exports[`flavors.basic validates required fields 2`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isArray": "requiredObjectArrayValue must be an array",
      "isDefined": "requiredObjectArrayValue should not be null or undefined",
    },
    "property": "requiredObjectArrayValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isArray": "requiredPrimitiveArrayValue must be an array",
      "isDefined": "requiredPrimitiveArrayValue should not be null or undefined",
    },
    "property": "requiredPrimitiveArrayValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "falseBooleanValue must be a boolean value",
      "isDefined": "falseBooleanValue should not be null or undefined",
    },
    "property": "falseBooleanValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "zeroBooleanValue must be a boolean value",
      "isDefined": "zeroBooleanValue should not be null or undefined",
    },
    "property": "zeroBooleanValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isBoolean": "requiredBooleanValue must be a boolean value",
      "isDefined": "requiredBooleanValue should not be null or undefined",
    },
    "property": "requiredBooleanValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredDateStringValue should not be null or undefined",
      "isIso8601": "requiredDateStringValue must be a valid ISO 8601 date string",
    },
    "property": "requiredDateStringValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDate": "requiredDateValue must be a Date instance",
      "isDefined": "requiredDateValue should not be null or undefined",
    },
    "property": "requiredDateValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredEnumValue should not be null or undefined",
      "isEnum": "requiredEnumValue must be a valid enum value",
    },
    "property": "requiredEnumValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredIntegerValue should not be null or undefined",
      "isInt": "requiredIntegerValue must be an integer number",
    },
    "property": "requiredIntegerValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredNestedValue should not be null or undefined",
    },
    "property": "requiredNestedValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredNumberValue should not be null or undefined",
      "isNumber": "requiredNumberValue must be a number conforming to the specified constraints",
    },
    "property": "requiredNumberValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredStringValue should not be null or undefined",
      "isString": "requiredStringValue must be a string",
    },
    "property": "requiredStringValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "requiredUUIDValue should not be null or undefined",
      "isUuid": "requiredUUIDValue must be a UUID",
    },
    "property": "requiredUUIDValue",
    "target": Example {},
    "value": undefined,
  },
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "stringValueWithDifferentWireName should not be null or undefined",
      "isString": "stringValueWithDifferentWireName must be a string",
    },
    "property": "stringValueWithDifferentWireName",
    "target": Example {},
    "value": undefined,
  },
]
`;
